### PR TITLE
Move new interface configuration to the instance settings handler

### PR DESCRIPTION
**Pull request by luis4a0**
The interfaces added with `set` were added in two stages: one in the instance settings handler, and one at `start`. This PR moves all the code to the instance settings handler, so there are no half states when snapshotting. It is based on the code to configure interfaces via cloud-init.

### CODE REVIEW

File: src/daemon/daemon.cpp
Comment (georgeliao) [inline src/daemon/daemon.cpp:3613]: Maybe change the code block line 3612 to 3624 to 
```
    if (const auto info = std::find_if(host_nets.cbegin(), host_nets.cend(), [br_interface](const auto& i) {
            return i.id == br_interface;
        }); info == host_nets.cend())
    {
        throw std::runtime_error(fmt::format(invalid_network_template, br_interface, mp::bridged_interface_key));
    }
    else if (info->needs_authorization && !user_authorized_bridges.count(br_interface))
    {
        throw mp::NonAuthorizedBridgeSettingsException("Cannot update instance settings", instance_name, br_interface);
    }
```
Using if initialization to localize the variable info and also changed `const auto& info` to `const auto info` since std::find_if return iterator by value.

File: src/daemon/daemon.cpp
Comment (georgeliao) [inline src/daemon/daemon.cpp:3648]: The name of the function `add_extra_interfaces_to_cloud_init` is not clear enough. It is changed to `apply_extra_interfaces_to_cloud_init` in the PR above this one. 

It basically will overwrite the whole network-config file in cloud-init, which also means the input should the whole vector of existing extra interfaces. So the right call should be `instance->add_extra_interfaces_to_cloud_init(spec.default_mac_address, spec.extra_interfaces);`. Having a function which adds one single interface does make sense. However, when I was writing the function, I took a shortcut of reusing the function `make_cloud_init_network_config`, which eventually leads to the function `add_extra_interfaces_to_cloud_init` takes the whole extra interfaces vector. Anyway, for now we can apply the quick fix and keep it like this and see later on if we should change to single network interface addition.

**Comment by georgeliao** [comment]
Hi @luis4a0 , thanks for the good work. It looks good to me in general. 

I left a few small comments to address. After that, I think we can merge.

File: src/daemon/daemon.cpp
Comment (luis4a0) [reply src/daemon/daemon.cpp]: Right, looks better the way you propose. Changed.

File: src/daemon/daemon.cpp
Comment (luis4a0) [reply src/daemon/daemon.cpp:3646]: Yeah, this sounds good to me. Thanks!

Comment (luis4a0) [reply src/daemon/daemon.cpp:3645]: Changed.

File: src/platform/backends/shared/base_virtual_machine.h
Comment (luis4a0) [reply src/platform/backends/shared/base_virtual_machine.h:56]: These two are functions in the `VirtualMachine` class. LXD is the only backend which implements cloud-init handling differently from the base VM. On the other hand, each backend implements adding networks interface in a different way. If we do merge both, we'll have to repeat code to add to the cloud-init in all the backends. We can find a way to refactor this in order to write less code, but I think it's not worth the effort: code is easier to read this way IMHO.

@georgeliao WDYT?

Comment (georgeliao) [reply src/platform/backends/shared/base_virtual_machine.h:56]: @ricab 
The `apply_extra_interfaces_to_cloud_init` function started as a function which applies **all the extra interfaces and the default address** to cloud-init. So it not easy to merge into one since the former `add_network_interface` is about adding **one** network interface into the backend. The reason `apply_extra_interfaces_to_cloud_init` started like this is because: 
1. It is based on the utility function `make_cloud_init_network_config` which was initially made for creating netowk-config file based on **all** the information. It was a shortcut from my side. 
2. The original code of the `add_bridged_interface` function has two steps in which the first loop call `add_network_interface` optionally and the second one just add everything into run_to_boot. So the switch to applying all extra interfaces to cloud-init seemed to fit that framework. 

Now, based on the current code, it makes sense to have a utility function to add one single extra interface into cloud-init and further merge the two functions into one. Maybe I can do this refactoring in the new PR (tracking instance id).

**Review by georgeliao — APPROVED** [review]

File: src/platform/backends/shared/base_virtual_machine.h
Comment (georgeliao) [reply src/platform/backends/shared/base_virtual_machine.h:56]: @ricab @luis4a0 I will merge like this for now. This small refactor can be addressed in the new PR.

Comment (ricab) [reply src/platform/backends/shared/base_virtual_machine.h:56]: @georgeliao, @luis4a0, I removed this from the merge queue, for now. We are a team, our work is open to discussion as a team, and I would like to reach a conclusion here before merging, if you don't mind. 

To clarify, I did not mean that everything should be put in a single block of code. I think it is good that there are two functions. Just not that they are both public. To the client of the VM, `apply_extra_interfaces_to_cloud_init` is just an implementation detail of `add_network_interface` (IIUC, there is no reason to call one without the other ?). If that is the case, we should prevent other code from depending on it.

I understand the development process that led to the 2 functions. But this PR is supposed to finish the transition from the original implementation by joining the two steps into a single one. 

If that requires a lot of effort, that needs to be taken into account. But I would like to understand what that effort is. As I see it, you just make the cloud-init function `protected`, rename a method and add another:

```
BaseVM::add_network_interface(...)
{
    add_network_interface_host(...); // renamed in descending VMs
    apply_extra_interfaces_to_cloud_init(...); // nothing else to change here
}
```

The tests can stay mostly the same too, because you can mock both steps and make `apply_extra_interfaces_to_cloud_init` accessible in the Stub/Mock VM.

If this adds other difficulties (e.g. juggling branches), or if there are other arguments that I am not seeing, I am entirely open to considering and debating them, possibly leaving this task for another PR. It is not a very big deal even. However, as a matter of principle, our work needs to remain open to peer review by others in the team.

Comment (luis4a0) [reply src/platform/backends/shared/base_virtual_machine.h:56]: @ricab I understand your point. Since this is a "stack" of PR's, and the part of the code you are referring to was already reviewed, I think what @georgeliao proposes to do is completely reasonable, and he added that he can address that on a new branch. Let me add that working in a branch for long time can be painful, it has proven to be better to merge small pieces of code one at a time. This particular branch had grown a lot and I prefer to merge it as soon as possible, given that it is working as intended and I would prefer to base my work (legacy snapshots) on main instead of a yet unmerged branch.

Comment (georgeliao) [reply src/platform/backends/shared/base_virtual_machine.h:56]: > I understand the development process that led to the 2 functions. But this PR is supposed to finish the transition from the original implementation by joining the two steps into a single one.

Sorry I was not very clear about this. The 2 steps this PR aimed at was coarse grain level where the empty string was exposed, that was already merged to one that is this function. The 2 steps I mentioned here are the sub-steps of the original 2nd step. 

As to whether we should merge these two into one, well, it can be done, `apply_extra_interfaces_to_cloud_init` can be a private function which only gets called as a sub-function of `BaseVM::add_network_interface` (new function) and `BaseVirtualMachine::restore_snapshot`. 

However, since I had doubts of the genericity of the `apply_extra_interfaces_to_cloud_init`, what I mean by this is that maybe a single purposed function would be better `add_one_extra_interface_to_cloud_init(const std::string& default_mac_addr, const NetworkInterface& extra_interface)` and let `apply_extra_interfaces_to_cloud_init` only get called by `BaseVirtualMachine::restore_snapshot`, as a result I would prefer defer this. 

I hope this clarify more.

Comment (ricab) [reply src/platform/backends/shared/base_virtual_machine.h:56]: @georgeliao I think I get what you are saying, but I still think we need a solution. As it stands, `add_network_interface()` alone will still leave the VM in a half state, correct? *Some* other method needs to be called to do the cloud-init part, regardless of whether it is the same that is used to restore the snapshot, right? And that method also has no use on its own, outside of the VM, other than to add an interface, correct? If so, then it should not be public.

I agree that it would be better not to require redundant information (existing interfaces), so I understand if you prefer to do those two things together. And @luis4a0, I understand your concern with long-lived branches. That is why I have supported merging multiple individual PRs for the bridging stuff, even if they are only part of the story. But the understanding is that the task is unfinished, so being merged does not mean it is set in stone. Some things may still need to change: before this PR, two methods were necessary because of the two steps, so it is normal that they would pass review; with this PR, that will no longer be the case.

So, I am OK with merging this one if that simplifies your live, but I would like to see a commitment that this will get sorted next and not forgotten. Are you guys OK with that?

Comment (georgeliao) [reply src/platform/backends/shared/base_virtual_machine.h:56]: @ricab 
I definitely can address this first in my next PR. Likely, two things in one. 
1. make `apply_extra_interfaces_to_cloud_init` private and it only gets called by `BaseVirtualMachine::restore_snapshot`. 
2. create `add_one_extra_interface_to_cloud_init` private function and it only gets called by `add_network_interface`.

Comment (ricab) [reply src/platform/backends/shared/base_virtual_machine.h:56]: OK, I am happy with that. Thank you @georgeliao.

Adding it back to the merge queue then.

**Pull request merged**
